src/ucs/datastruct/ptr_array.h

```diff
@@ -121,12 +135,133 @@ void *ucs_ptr_array_replace(ucs_ptr_array_t *ptr_array, unsigned index, void *ne
-    for (_index = 0; _index < (_ptr_array)->size; ++_index) \
-         if (!__ucs_ptr_array_is_free(_var = (void*)((_ptr_array)->start[_index]))) \
+    uint32_t step_size = 0;\
+    for (_index = 0; (_index < (_ptr_array)->size); _index += step_size)\
+       if ((ucs_likely(!__ucs_ptr_array_is_free_and_get_step_size((ucs_ptr_array_elem_t)(_var = (void *)((_ptr_array)->start[_index])), &step_size))))\
+
+
+/*
+   Locked interface
+*/
+
+/* Locked ptr array */
+typedef struct ucs_ptr_array_locked {
+    ucs_ptr_array_t super;
+    ucs_recursive_spinlock_t lock;
+} ucs_ptr_array_locked_t;
+
+/**
+ * Locked array init
+ */
+void ucs_ptr_array_locked_init(ucs_ptr_array_locked_t *locked_ptr_array, uint32_t init_placeholder,
+                        const char *name);
+
+
+/**
+ * Cleanup the locked array.
+ *
+ * @param locked_ptr_array    Pointer a locked ptr array.
+ *
+ * All values should already be removed from it.
+ */
+void ucs_ptr_array_locked_cleanup(ucs_ptr_array_locked_t *locked_ptr_array);
+
+
+/**
+ * Insert a pointer to the locked array.
+ *
+ * @param value        Pointer to insert. Must be 8-byte aligned.
+ * @param placeholder  Filled with placeholder value.
+ * @return             The index to which the value was inserted.
+ *
+ * Complexity: amortized O(1)
+ *
+ * Note: The array will grow if needed.
+ */
+unsigned ucs_ptr_array_locked_insert(ucs_ptr_array_locked_t *locked_ptr_array, void *value,         
+                   uint32_t *placeholder_p);
 
 
+/**
+ * Remove a pointer from the locked array.
+ *
+ * @param locked_ptr_array  Pointer a locked ptr array.
+ * @param index             Index to remove from.
+ * @param placeholder       Value to put in the free slot.
+ *
+ * Complexity: O(1)
+ */
+void ucs_ptr_array_locked_remove(ucs_ptr_array_locked_t *locked_ptr_array, unsigned index,
+                          uint32_t placeholder);
+
+
+/**
+ * Replace pointer in the locked array
+ *
+ * @param  index    index of slot
+ * @param  new_val  value to put into slot given by index
+ * @return old value of the slot
+ *
+ * Complexity: O(1)
+ */
+void *ucs_ptr_array_locked_replace(ucs_ptr_array_locked_t *locked_ptr_array,
+	unsigned index, void *new_val);
+
+
+/* Acquire ptr_array lock */
+#define ucs_ptr_array_locked_acquire_lock(_locked_ptr_array)\
+    ucs_recursive_spin_lock(&(_locked_ptr_array)->lock)
+
+
+/* Release ptr_array lock */
+#define ucs_ptr_array_locked_release_lock(_locked_ptr_array)\
+    ucs_recursive_spin_unlock(&(_locked_ptr_array)->lock)
+
+
+/**
+ * Retrieve a value from the locked array.
+ *
+ * The user is required to lock using ucs_ptr_array_locked_acquire_lock()
+ * and ucs_ptr_array_locked_release_lock() functions.
+ *
+ * @param index   Index to retrieve the value from.
+ * @param value   Filled with the value.
+ * @return        Whether the value is present and valid.
+ *
+ * Complexity: O(1)
+ */
+#define ucs_ptr_array_locked_lookup(_locked_ptr_array, _index, _var)\
```

Comment:
maybe worth adding lock/unlock inside? like with other locked_array API

Reply:
Well, 2 points here:
1) Due to the usage model of the previously defined macro ucs_ptr_array_lookup() - It needs to set the _var value and it's used throughout the code as a definition inside a a if() statement etc.
Therefore, the following code will not work:
#define ucs_ptr_array_locked_lookup(_locked_ptr_array, _index, _var) { \
    ucs_ptr_array_locked_acquire_lock(_locked_ptr_array);\
    ucs_ptr_array_lookup(&(_locked_ptr_array)->super, _index, _var);\
    ucs_ptr_array_locked_release_lock(_locked_ptr_array); }

2) That said, we still need the ucs_ptr_array_locked_lookup() function since the ptr_array pointer is pointed to via locked->super ==> We don't expect the user to look into the internal structure of the ptr array.

3) Still, as you say, the lookup() function is not locked indeed - So the worst case scenario, since this is not a function that updates the ptr_array, is that the lookup will bring the previous pointer from the array before it was updated / initialized by insert() or replace().

4) We might want to consider renaming the macro to reflect that the user must wrap it with lock() / unlock() - but I can't think of any good name at the moment :-)
Please notice, that I wrote in the macro prototype that the user is required to lock/unlock when using this macro if needed in his use-case.
/**
 * Retrieves a value from the locked array.
 *
 * The user is required to lock using ucs_ptr_array_locked_acquire_lock()
 * and ucs_ptr_array_locked_release_lock() functions.
 *
 * @param [in] locked_ptr_array  Pointer to a locked ptr array.
 * @param [in] index             Index to retrieve the value from.
 * @param [out] _var             Filled with the value.
 * @return                       Whether the value is present and valid.
 *
 * Complexity: O(1)
 */

@brminich , Please let me know what you think or if you have a recommendation how to implement the locked_lookup() macro...

Reply:
i did not get these points. You protect the array itself, not the out value, right?
BTW, can you please share use case it is intended for?
Are you using fine-grained multithreading so that you have to protect an array separately?

Reply:
i did not get these points. You protect the array itself, not the out value <-- Indeed.

The use-case is when we need a thread-safe ptr_array that is being updated by multiple threads.

Reply:
@brminich 

I asked shuki to add the locked-ptr-array API has several purposes:
1. as discussed with yossi in #4796, we want changes in ptr-array to effect ucs/async , especially that problematic timerq which calls realloc() on every new timer:
https://github.com/openucx/ucx/blob/master/src/ucs/time/timerq.c#L44
(thanks @tanyabrokhman for pointing this out)
My intention is to rewrite timerq on top of ptr-array, to improve performance.

2. I need to use pointer-array directly for collectives, where locking makes a lot of sense.

In general, I think a "thread-safe" ptr-array makes a lot of sense, and besides my two use cases I'm sure if I look for UCX calls to realloc() - I'll find more :)

Specifically about the value I think we should provide a coherent API for the locked_ptr_array, so that there's no need to mix "locked" and "regular" API. Regarding locking access specifically - this is meant to protect from "get()" during foreach, reallocation, and other lengthy operations.

Reply:
Ok, thanks for the info. Why not implementing locked_lookup with actual locks?

Reply:
@brminich you're right. I just talked to shuki about this - he's working on a proposal for that.

Reply:
@dmitrygx  - Please review the latest code.
Following a discussion with @alex--m , please see a nice and elegant solution for most concerns and comment.